src/MongoDB.Driver/Linq/Processors/ParameterExpressionCollector.cs

```diff
@@ -0,0 +1,53 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+
+namespace MongoDB.Driver.Linq.Processors
+{
+    /// <summary>
+    /// The expression visitor to collect all parameters in the expression.
+    /// </summary>
+    internal sealed class ParameterExpressionCollector : ExpressionVisitor
+    {
+        private readonly List<string> _paramsCollection = new List<string>();
+
+        /// <summary>
+        /// Collects the list of parameters which are used in the expression.
+        /// </summary>
+        /// <param name="expression">The expression.</param>
+        /// <returns>The result collection.</returns>
+        public static IEnumerable<string> Collect(Expression expression)
+        {
+            var parameterCollector = new ParameterExpressionCollector();
+            parameterCollector.Visit(expression);
+            return parameterCollector._paramsCollection.Distinct();
+        }
+
+        /// <inheritdoc />
+        protected override Expression VisitMember(MemberExpression node)
+        {
+            var parameterExpression = node.Expression as ParameterExpression;
```

Comment:
Isn't there a VisitParameter method to override?

Reply:
There is, but `VisitParameter` method also is triggered for the input expression parameters.
So, in this expression: 
` (a) => true`
`VisitParameter` will be triggered, but `VisitMember` not.

Reply:
Ok, makes sense. What about this `(a) => a`?  I don't think you'll end up collecting `a` because it's not part of a member expression.

Reply:
Good point, my implementation doesn't collect the parameter from your expression.

but I think this expression cannot be used in any LINQ conditions like `where` or `any`, so this problem will not affect this ticket implementation. But I guess, in this case, this class should be renamed or even placed as a nested visitor in `ExpressionValidation`.

However, do you know how can we get only parameters which are used in the expression (without input expression params)? As I see, there is no any property/flag in `ParameterExpression` to show the difference between input expression parameters and parameters which are used inside expression?

Reply:
If A is an array of booleans, `[true, false, true]`, then `myBooleanArray.Any(a => a)` should be valid I believe. Translated to `{myBooleanArray: { $elemMatch: { $in[true]}}}`.

Might be wrong here... it's been a while since I've dealt with elemMatch.

Reply:
`myBooleanArray.Any(a => a)`
my implementation collects `a` parameter:

        private class TestObject
        {
            public int Value { get; set; }
            public IEnumerable<bool> Collection { get; set; }
        }
        [Fact]
        public void BoolTest()
        {
            Expression<Func<TestObject, bool>> expr = (a) => a.Collection.Any(b => b);
            IEnumerable<string> result = ParameterExpressionCollector.Collect(expr);
            //result[0] = "a"
        }

Reply:
@craiggwilson, my previous comment(which I've deleted) was wrong. 
I've changed `ParameterExpressionCollector` and now I'm visiting **VisitParameter** instead of **VisitMember**.

Let me show the test which I've created:
`(a) => a.Collection.Any(b => b)`
where **Collection** is IEnumerable of `bool`
This test doesn't throw an error, but the generated query, in this case, is wrong:
`{ "Collection2" : { "$elemMatch" : { "" : true } } }`
@craiggwilson, @rstam, how do you think should we consider this behavior as a bug and create a new ticket for this?